### PR TITLE
Flush caches

### PR DESCRIPTION
**Pull request by jeabakker**
Introducing a general cache reset function. This function triggers an event to which all caches can register their reset functions.

This was implemented for:
- system_cache
- simplecache
- memcache

### CODE REVIEW

**Comment by hypeJunction** [comment]
Could the function maybe take in some params in case we want to filter which caches to flsuh?

**Comment by jeabakker** [comment]
how would we check that, when all the function does is trigger an event?

maybe $type = 'all' and then
trigger_event('cache:flush', $type)

so simplecache and system cache can register on $type = 'system'
and memcache register on $type = 'memcache'
and opcaches register on $type = 'opcache'

**Comment by jdalsem** [comment]
> Could the function maybe take in some params in case we want to filter which caches to flsuh?

if you want individual caches to be flushed, you can use those specific flush functions...

i think this is meant as an easy flush *all*

**Comment by hypeJunction** [comment]
Hmm. Maybe I am just overthinking this. Let's wait and see what others think. I don't think I know enough about our caches.

File: actions/admin/site/flush_cache.php [hunk 1]
```diff
@@ -3,8 +3,7 @@
  * Flush all the caches
  */
 
-elgg_invalidate_simplecache();
-elgg_reset_system_cache();
+elgg_flush_caches();
 _elgg_services()->autoloadManager->deleteCache();
```
Comment (jdalsem) [inline actions/admin/site/flush_cache.php]: shouldn't this also be moved to a event hook?

File: actions/admin/site/update_advanced.php [hunk 2]
```diff
@@ -85,7 +85,7 @@
 	// if you cancel this even you should present a message to the user
 	if (elgg_trigger_before_event('regenerate_site_secret', 'system')) {
 		init_site_secret();
-		elgg_reset_system_cache();
+		elgg_flush_caches();
```
Comment (jdalsem) [inline actions/admin/site/update_advanced.php]: why does all cache needs to be flushed here?

File: engine/lib/cache.php [hunk 3]
```diff
@@ -222,6 +222,16 @@ function elgg_invalidate_simplecache() {
 }
 
 /**
+ * Flush all the registered caches
+ * 
+ * @return void
+ * @since 1.11
+ */
+function elgg_flush_caches() {
+	_elgg_services()->events->trigger('cache:flush', 'system');
```
Comment (jdalsem) [inline engine/lib/cache.php]: this new event should probably be added to the RST docs

File: upgrade.php [hunk 4]
```diff
@@ -95,7 +95,7 @@
 	);
 
 	// reset cache to have latest translations available during upgrade
-	elgg_reset_system_cache();
+	elgg_flush_caches();
```
Comment (jdalsem) [inline upgrade.php]: no need to flush all... just reset system cache, as the comment states it is only needed for the translations

File: engine/lib/cache.php [hunk 5]
```diff
@@ -234,4 +244,15 @@ function _elgg_cache_init() {
...
+	// register plugin hooks for cache reset
+	$events->registerHandler('cache:flush', 'system', 'elgg_reset_system_cache');
+	$events->registerHandler('cache:flush', 'system', 'elgg_invalidate_simplecache');
+	
+	// register some PHP caches for reset
+	// @see http://php.net/manual/en/refs.basic.php.php
+	$events->registerHandler('cache:flush', 'system', 'apc_clear_cache');
+	$events->registerHandler('cache:flush', 'system', 'opcache_reset');
+	$events->registerHandler('cache:flush', 'system', 'wincache_refresh_if_changed');
+	$events->registerHandler('cache:flush', 'system', 'wincache_ucache_clear');
```
Comment (jdalsem) [inline engine/lib/cache.php]: maybe add a code comment that during the event handling it is checked if the function is callable..

File: actions/admin/site/flush_cache.php (hunk 1 above)
Comment (jeabakker) [inline actions/admin/site/flush_cache.php]: I tried to find a correct place to do this, but couldn't find one i was happy with. Suggestions?

File: engine/lib/cache.php (hunk 3 above)
Comment (jeabakker) [inline engine/lib/cache.php]: added the docs

**Comment by jdalsem** [comment]
LGTM

**Comment by hypeJunction** [comment]
Do we need to test anything here?

**Comment by jdalsem** [comment]
> Do we need to test anything here?

don't think so its just a bunch of event triggers.... this code LGTM

File: engine/classes/ElggMemcache.php [hunk 6]
```diff
@@ -107,6 +107,9 @@ public function __construct($namespace = 'default') {
 		if (isset($this->CONFIG->memcache_expires)) {
 			$this->expires = $this->CONFIG->memcache_expires;
 		}
+		
+		// make sure memcache is reset
+		_elgg_services()->events->registerHandler('cache:flush', 'system', array($this, 'clear'));
```
Comment (Srokap) [inline engine/classes/ElggMemcache.php]: As we here use method called `clear` I'd prefer event to be called `cache:clear` as well.

Comment (jeabakker) [inline engine/classes/ElggMemcache.php]: for ElggMemcache it's clear for Memcache it's actualy flush. for simplecache it's invalidate, for system_cache it's reset.....

so let's just stick with flush (as that is the action it was envisioned for)

File: engine/lib/cache.php [hunk 7]
```diff
@@ -222,10 +222,21 @@ function elgg_invalidate_simplecache() {
...
+function elgg_flush_caches() {
+	_elgg_services()->events->trigger('cache:flush', 'system');
+}
+
+/**
  * Initializes the simplecache lastcache variable and creates system cache files
  * when appropriate.
  * 
  * @access private
+ * @
```
Comment (mrclay) [inline engine/lib/cache.php]: typo?

File: engine/lib/cache.php [hunk 8]
```diff
@@ -234,4 +245,16 @@ function _elgg_cache_init() {
 
 return function(\Elgg\EventsService $events, \Elgg\HooksRegistrationService $hooks) {
 	$events->registerHandler('ready', 'system', '_elgg_cache_init');
+	
+	// register plugin hooks for cache reset
+	$events->registerHandler('cache:flush', 'system', 'elgg_reset_system_cache');
+	$events->registerHandler('cache:flush', 'system', 'elgg_invalidate_simplecache');
+	
+	// register some PHP caches for reset
+	// no need to check if the functions are callable, this is handled in the EventService
```
Comment (mrclay) [inline engine/lib/cache.php]: I don't think this is a good idea

Comment (mrclay) [inline engine/lib/cache.php]:
```php
foreach (['apc_clear_cache', ...] as $func) {
    if (is_callable($func)) {
```

**Comment by mrclay** [comment]
I don't like assuming these external caches we don't use should be cleared. The hook makes it easy enough for plugins to manually register for `cache:flush` if they want to do this.

File: engine/lib/cache.php (hunk 8 above)
Comment (jdalsem) [inline engine/lib/cache.php]: the is_callable check is already done in the event_handler

**Comment by jdalsem** [comment]
> I don't like assuming these external caches we don't use should be cleared. The hook makes it easy enough for plugins to manually register for cache:flush if they want to do this.

disagree... this PR allows those caches to be flushed automatically when you have the related modules enabled. They are the standard modules from PHP where elgg runs on, so it makes a lot of sense to flush them. The alternative is that if you have Elgg and use APC and you need to flush the APC cache, you need to program just for that, i do not think that is very nice to the sysadmins/community masters.

**Comment by juho-jaakkola** [comment]
Doesn't it cause trouble that opcode caches are flushed on the same request as the others? Wouldn't this mean the caches need to be flushed twice by the admin?
 1. request: The opcode cache gets flushed, but other caches are still using the old opcode cache
 2. request: Other caches can now regenerate using the updated code in the opcode cache

**Comment by jeabakker** [comment]
> Doesn't it cause trouble that opcode caches are flushed on the same request as the others?

not as far as we experienced. We have both apc and opcache on some systems and with the reset functions we never had any trouble

**Comment by mrclay** [comment]
Elgg is frequently not the only PHP app on a system, and we don't know how other apps will be affected. Wouldn't you consider this poor form if WordPress cleared caches Elgg was using? Again, it's trivial to wire these functions up in a plugin.

**Comment by jeabakker** [comment]
updated the pr. The opcaches are no longer reset. But since there is a hook for it a plugin can be made to handle the reset

File: docs/guides/events-list.rst [hunk 9]
```diff
@@ -61,6 +61,9 @@ System events
 **init:cookie, <name>**
     Return false to override setting a cookie.
 
+**cache:flush, system**
+    Reset internal and external caches. For example it resets system_cache and simplecache, but also PHP opcode caches (APC, OPCache, WinCache).
```
Comment (mrclay) [inline docs/guides/events-list.rst]: Maybe: `Reset internal and external caches, by default including system_cache, simplecache, and memcache. One might use it to reset others such as APC, OPCache, or WinCache.`

**Comment by jeabakker** [comment]
updated docs with @mrclay suggestion

**Comment by mrclay** [comment]
Rebased in #8308

**Pull request closed without merging**